Added more version comparison operations

nivedwho opened the pull request:
Currently the [TorchVersion](https://github.com/pytorch/pytorch/blob/1022443168b5fad55bbd03d087abf574c9d2e9df/torch/torch_version.py#L13) only only supports 'greater than', and 'equal to' operations for comparing torch versions and something like `TorchVersion('1.5.0') < (1,5,1)` or `TorchVersion('1.5.0') >= (1,5)` will throw an error.

I have added 'less than' (`__lt__()`), 'greater than or equal to' (`__ge__()`) and 'less than or equal to' (`__le__()`) operations, so that the TorchVersion object can be useful for wider range of version comparisons.     

cc @seemethere @zsol

nivedwho committed:
add comparison operations

Added 'less than', 'greater than or equal to' and 'less than or equal to' operations for comparing versions.

nivedwho committed:
correct spacing

nivedwho commented:
@seemethere can you review this PR?

seemethere commented:
This shouldn't necessarily be the case since we're using the `@total_ordering` decorator `functools`, see: https://docs.python.org/3/library/functools.html#functools.total_ordering

I wonder why it is producing an error here when it should obviously support all of these operators

seemethere reviewed (approved):
We should also probably remove the `total_ordering` decorator as well since it's not actually used and seems to be bugged for doing comparisons against multiple types.

nivedwho committed:
remove total_ordering decorator

nivedwho committed:
Lint

nivedwho committed:
fixed total_ordering issue

Currently the `total_ordering` decorator is not working, hence other comparison operations are not possible. To solve this issue. a base class is introduced that is not directly inherited from the input string. This way `@total_ordering` works and all the comparison operations are supported.

nivedwho committed:
Added comparison operations

`@total_ordering` seems to be not working for comparing multiple types, since `TorchVersion` class is directly inherited from the input string. 
Hence for supporting multiple comparison operations, 'less than', 'greater than or equal to' and 'less than or equal to' operations are added and the decorator is removed.

nivedwho commented:
@seemethere how can I get this PR merged? Is there anything left for me to do?

seemethere commented:
> @seemethere how can I get this PR merged? Is there anything left for me to do?

I've run into some internal issues with merging this, going to try and rebase it to see if that resolves

seemethere committed:
get rid of extra __ge__, re-order them to make more sense

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>
